### PR TITLE
PVR documentation change for specials

### PR DESCRIPTION
**Pull request by emveepee**
## Description

Documentation was stating the PVR addons could use iSeriesNumber=iEpisodeNumber=0  which is no longer acceptable

Documentation change only to reflect comment from @ksooo on PR17408 https://github.com/xbmc/xbmc/pull/17408#issuecomment-597760690 

## Motivation and Context

Confusion for PVR addon developers.   

Many pvr addons may need to be updated of kodi API design change this PR documents the requirement

## How Has This Been Tested?

Cannot easily test.  Should be populated to codedoc.xyz after release

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ x] **None of the above** (please explain below)

Documentation change only however from a pvr addon perspective the change is a cosmetic change.

## Checklist:

- [ x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ x] My change requires a change to the documentation, either Doxygen or wiki
- [ x] I have updated the documentation accordingly
- [ x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by ronie — APPROVED** [review]

**Review by phunkyfish — APPROVED** [review]
LGTM

**Comment by phunkyfish** [comment]
Should we also change these docs for EPG Entries? They appear to suffer from a similar issue and are assumed to be specials if season is 0.

**Comment by ksooo** [comment]
Can you do me a favor and update also the documentation for the respective fields of EPG_TAG too match PVR_RECORDING?

**Comment by ksooo** [comment]
@phunkyfish you beat me to this. 😉

**Comment by phunkyfish** [comment]
There’s a first time for everything...

**Comment by emveepee** [comment]
I specifically didn't do the this for EPG because that is a code change.

https://github.com/xbmc/xbmc/blob/dfad22ee2edc7f5ba6708c2a4e38d002f7629987/xbmc/pvr/guilib/guiinfo/PVRGUIInfo.cpp#L567

Also S0 for specials is a hack.  The Video Library with filename lookups can use TVDB's season 0 database info which is probably where it started, but in EPG you break all addons for no gain.  Specials should also be shown in the guide for at a glance view so it should be treated like the New flag.

**Comment by ksooo** [comment]
No gain? I see this a bit different. Gain is a clean, understandable and consistent API. This is worth a lot.

**Comment by phunkyfish** [comment]
But what does the calling code do with the false return? Does it just assume it's zero?

**Comment by ksooo** [comment]
Return false means that the GUI info provider did not handle the requested GUI info label/bool/int and next provider should be asked. GUI info providers are chained.

**Comment by ksooo** [comment]
For me, all this code freaking with season and episode numbers in pvrguiinfo looks like a mess. Clearly a candidate for fixes/cleanup.

**Comment by ksooo** [comment]
True/false return conditions seen to be wrong at several places in that file. I must have been drunk when I was writing that.

**Comment by emveepee** [comment]
Yes I was curious to see what S0 E1 would look like after any change something presumably like 
S0 ES01

Even then specials will look ugly anyway with just the S0 as I posted yesterday https://github.com/xbmc/xbmc/pull/17408#issuecomment-597376079

**Comment by phunkyfish** [comment]
Ok, but if I set season to zero and episode to a number in an addon it will display as ES7 for example. It must infer that logic somewhere.

**Comment by ksooo** [comment]
> it will display as ES7 for example. It must infer that logic somewhere.

 is it in the Estuary skin xml?

**Comment by emveepee** [comment]
@ronie explained the old logic on the PR that started this discussion https://github.com/xbmc/xbmc/pull/17408 and the episode number with the leadin S was how it worked before.  You may not have been drunk

**Comment by phunkyfish** [comment]
@emveepee if you just change that one line to >= 0  for series number what’s the impact?

**Comment by phunkyfish** [comment]
Could be @ksooo will look when next at a PC.

**Comment by emveepee** [comment]
The >= change for the EPG will mean all PVR addons need to be updated to default iSeriesNumber to -1 or they will be considered specials

**Comment by phunkyfish** [comment]
But they already are right? For me they are all start ES with no code change.

**Comment by emveepee** [comment]
>
>But they already are right? For me they are all start ES with no code change.

I didn't compile it but I see what you mean S0 with E>0 is always odd so it is ES7 and with the change it could be S0 ES7  Since no EPG source I know would do this or match TVDB  ie. https://www.thetvdb.com/series/the-big-bang-theory/seasons/official/0 for specials from a PVR perspective user would never see it without a glitch in the guide date

**Comment by phunkyfish** [comment]
Here is what I found in the skin:
```
	<variable name="SeasonEpisodeLabel">
		<value condition="String.IsEmpty(ListItem.EpisodeName)">$INFO[ListItem.Season,S]$INFO[ListItem.Episode,E]</value>
		<value>$INFO[ListItem.Season,S]$INFO[ListItem.Episode,E,: ]</value>
	</variable>
```

Not sure how that becomes ES though.

**Comment by emveepee** [comment]
>
>Not sure how that becomes ES though.

The skin adds the E to the S7 that comes from https://github.com/xbmc/xbmc/blob/dfad22ee2edc7f5ba6708c2a4e38d002f7629987/xbmc/pvr/guilib/guiinfo/PVRGUIInfo.cpp#L578

**Comment by phunkyfish** [comment]
So this means there is no code change needed and it's safe to change the docs to EPG_TAG also.

**Comment by emveepee** [comment]
> 
> 
> So this means there is no code change needed and it's safe to change the docs to EPG_TAG also.
> 
> The docs were just missing for EPG_TAG all along as it already had the correct behaviour.

No it isn't because the current code only works when the iEpisodeNumber is > 0  and I don't know if the S should be prepadded.  There is nothing documented that says it can't be invalid or zero at least as going by the Video Library https://github.com/xbmc/xbmc/blob/dfad22ee2edc7f5ba6708c2a4e38d002f7629987/xbmc/guilib/guiinfo/VideoGUIInfo.cpp#L221   (which also needs a cleanup)

**Comment by phunkyfish** [comment]
I guess based on the other PR the S should be dropped and bring it inline with how recordings work also.

**Comment by phunkyfish** [comment]
The following code should bring it inline with how it was broken for Recordings consistently across all addons. So a special would now be S0E1 for example, the same for both.

In PVRGUIInfo.cpp:

```
      case LISTITEM_SEASON:
        if (epgTag->SeriesNumber() >= 0)
        {
          strValue = StringUtils::Format("%i", epgTag->SeriesNumber());
          return true;
        }
        return false;
      case VIDEOPLAYER_EPISODE:
      case LISTITEM_EPISODE:
        if (epgTag->EpisodeNumber() >= 0)
        {
          strValue = StringUtils::Format("%i", epgTag->EpisodeNumber());
          return true;
        }
        return false;
```

And in EpgInfoTag:

```
    int m_iSeriesNumber = -1; /*!< series number */
    int m_iEpisodeNumber = -1; /*!< episode number */
    int m_iEpisodePart = -1; /*!< episode part number */
```

**Comment by ronie** [comment]
@phunkyfish i've already created a PR for it,
please review and let me know if anything is missing.

https://github.com/xbmc/xbmc/pull/17488

**Comment by phunkyfish** [comment]
@emveepee in light of the other PR can you update the docs for EPG_TAG too?

**Comment by emveepee** [comment]
> 
> 
> @emveepee in light of the other PR can you update the docs for EPG_TAG too?

Changed.   Dev's using PVR_TIMER_TYPE_REQUIRES_EPG_SERIES_ON_CREATE might need to consider S0 since users might consider a special part of a series.     Since there is no S0 for specials in the wild for EPG data (except possibly pirate VOD)  it is a trivial point.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h
Comment (phunkyfish) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: Can you also add for part number?

**Comment by phunkyfish** [comment]
> > @emveepee in light of the other PR can you update the docs for EPG_TAG too?
> 
> Changed. Dev's using PVR_TIMER_TYPE_REQUIRES_EPG_SERIES_ON_CREATE might need to consider S0 since users might consider a special part of a series. Since there is no S0 for specials in the wild for EPG data (except possibly pirate VOD) it is a trivial point.

Should probably update the docs there too so they are all >= 0. Just trying to find where that impacts the code.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h
Comment (emveepee) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: Where is iEpisodePartNumber documented that 0 isn't invalid?   Also I don't think it is even displayed and the tests on it for series test are all > 0

I do have reservations that  iEpisodeNumber needs to -1 since this should be reserved for values where the guilib functions should display it and it doesn't with 0.

Comment (phunkyfish) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: It is not documented like that yet. This is what we are defining. In the other PR I have asked Ronie to update the series test also to match these values.

The idea with @ronie’s original PR is that it should be possible to set the values to any positive integer. Which makes sense. I have seen many pilots there are episode 0, e.g. S01E00.

Comment (phunkyfish) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: Ronie will make a change to the other PR so episode will be `>= 0`.

Comment (emveepee) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: Pilots as S>0E00 sounds like an afterthought on the TVDB and I have never seen it in the wild on TV.  In any case I was  ignoring that discussion   It is less clear why iEpisodePart number needs to be -1  I have often seen EPG with parts 1/3, 1/8 etec but  what is the use case for 0/5?  (the limit isn't part of Kodi spec but I put it in for this question.

Comment (phunkyfish) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: Episode part number is supposed to be used for when a single episode get's split for some reason. I.e. and news programme in between. Depending on your country they may be use part 1 and 2 or they may use part 0 and 1. 

We don't use this currently in Kodi but in the future if we did -1 for invalid would be correct.

Comment (emveepee) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: I have had been told not to commit for future undocumented possibilities and since your use case would likely require a code update https://github.com/xbmc/xbmc/blob/master/xbmc/pvr/epg/EpgInfoTag.cpp#L646 I still am reluctant to include it as documentation

Comment (phunkyfish) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: The code exists already. If we are going to change every PVR addon to support the new invalid values I would rather do it once ;)

I have asked Ronie to update EpgInfoTag L646.

Comment (emveepee) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: Ok I will make the documentation chance but you have turned this from a cosmetic change to a breaking change since all the PVR addons  that haven't changed to -1 will tag all EPG items as IsSeries() true

Comment (phunkyfish) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: Yes, but I will update all the addons to set defaults as -1. Can merge this and Ronnie's PR once I have them ready.

Comment (emveepee) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: The issue isn't the default.  I looked at this when @ksooo said he didn't have time and if you don't know whether the backend supports E0  Using pvr.vuplus as an example https://github.com/kodi-pvr/pvr.vuplus/blob/Matrix/src/enigma2/data/RecordingEntry.cpp#L204

  left.iSeriesNumber = m_seasonNumber;
  left.iEpisodeNumber = m_episodeNumber;

you need to add

if   ( left.iSeriesNumber == 0 && left.iEpisodeNumber ==0 )
{
  left.iSeriesNumber = -1;
  left.iEpisodeNumber = -1;
}

That  still leaves open the use case that you suggested with the >= testing that the E0 could be valid now so S0 E0 could be a special.  Only the maintainer would know that.

Comment (phunkyfish) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: I don't follow. In this case you would just set the default in BaseEntry.h to -1 instead of 0.

Comment (emveepee) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: And what does that do to https://github.com/kodi-pvr/pvr.vuplus/blob/0a9d2c0efdd336e579dc988d171a20244b666632/src/enigma2/extract/ShowInfoExtractor.cpp#L63  and 77

That one will be easier then others too since we can see how they are formed.  If this is all done on the backend you don't

Comment (phunkyfish) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: If what you mean is that the solution is different in each addon, agreed.

Comment (phunkyfish) [reply xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_epg_types.h]: Here's the PR for vuplus: https://github.com/kodi-pvr/pvr.vuplus/pull/275

As each PR is created the maintainer can raise if a different change is needed. I agree there will be addons that require more work. But this needs to be done regardless.

**Comment by phunkyfish** [comment]
Ok, I'm going to merge this first as I would like to add a define to pvr_epg_types before finishing the addon PRs.

**Comment by phunkyfish** [comment]
Thanks @emveepee for bearing with me, and for the contribution ;)

**Comment by phunkyfish** [comment]
WIN32 failure unrelated

**Pull request merged**
